drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,955 @@
+		return -rte_errno;
+	}
+
+	//TO_DO: need to confirm following:
+	priv->vfio_dev_fd = -1;
+	priv->vfio_group_fd = -1;
+	priv->vfio_container_fd = -1;
+
+	rte_pci_device_name(&pci_dev->addr, devname, RTE_DEV_NAME_MAX_LEN);
+	ret = rte_vfio_get_group_num(rte_pci_get_sysfs_path(), devname,
+			&iommu_group_num);
+	if (ret <= 0) {
+		DRV_LOG(ERR, "%s failed to get IOMMU group", devname);
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	priv->vfio_container_fd = rte_vfio_container_create();
+	if (priv->vfio_container_fd < 0) {
+		DRV_LOG(ERR, "%s failed to get container fd", devname);
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	priv->vfio_group_fd = rte_vfio_container_group_bind(
+			priv->vfio_container_fd, iommu_group_num);
+	if (priv->vfio_group_fd < 0) {
+		DRV_LOG(ERR, "%s failed to get group fd", devname);
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	priv->pdev = pci_dev;
+
+	priv->vpdev = virtio_pci_dev_alloc(pci_dev);
+	if (priv->vpdev == NULL) {
+		DRV_LOG(ERR, "%s failed to alloc virito pci dev", devname);
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	priv->vfio_dev_fd = rte_intr_dev_fd_get(pci_dev->intr_handle);
+	if (priv->vfio_dev_fd < 0) {
+		DRV_LOG(ERR, "%s failed to get vfio dev fd", devname);
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	priv->vdev = rte_vdpa_register_device(&pci_dev->device, &virtio_vdpa_ops);
+	if (priv->vdev == NULL) {
+		DRV_LOG(ERR, "Failed to register vDPA device.");
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	priv->hw_nr_virtqs = virtio_pci_dev_nr_vq_get(priv->vpdev);
+	ret = virtio_vdpa_queues_alloc(priv);
+	if (ret) {
+		DRV_LOG(ERR, "Failed to alloc vDPA device queues.");
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	priv->nvec = virtio_pci_dev_interrupts_num_get(priv->vpdev);
+	if (priv->nvec <= 0) {
+		DRV_LOG(ERR, "Error dev interrupts %d less than 0", priv->nvec);
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	ret = virtio_pci_dev_interrupts_alloc(priv->vpdev, priv->nvec);
+	if (ret) {
+		DRV_LOG(ERR, "Error alloc virtio dev interrupts: %s",
+				strerror(errno));
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	pthread_mutex_init(&priv->vq_config_lock, NULL);
+	pthread_mutex_lock(&priv_list_lock);
+	TAILQ_INSERT_TAIL(&virtio_priv_list, priv, next);
+	pthread_mutex_unlock(&priv_list_lock);
+	return 0;
+
+error:
+	if (priv) {
+		rte_free(priv);
+	}
+	return -rte_errno;
+}
+
+static int
+virtio_vdpa_dev_remove(struct rte_pci_device *pci_dev)
+{
+	struct virtio_vdpa_priv *priv = NULL;
+	int found = 0,ret;
+
+	pthread_mutex_lock(&priv_list_lock);
+	TAILQ_FOREACH(priv, &virtio_priv_list, next) {
+		if (priv->pdev == pci_dev) {
+			found = 1;
+			break;
+		}
+	}
+	if (found)
+		TAILQ_REMOVE(&virtio_priv_list, priv, next);
+	pthread_mutex_unlock(&priv_list_lock);
+	if (found) {
```

Comment:
write this as,

if (!found)
  return -ENODEV:

Code of line 917 to .. comes here.